fix(client-fetch): ensures plain/text body is sent with request

franworks opened the pull request:
resolves: #2553

When the `options.serializedBody` property was introduced, it replaced the body sent in the Request object causing any `options.body` not requiring a `bodySerializer` (ie plain/text) to be omitted from the request.

This PR fixes this behavior, ensuring that when an `options.body` is defined it is sent with the Request 'as-is', or 'serialized' if an `options.bodySerializer` is defined.

Commit:
fix(client-fetch): ensures plain/text body is sent with request

resolves: #2553

mrlubos commented on packages/openapi-ts/src/plugins/@hey-api/client-fetch/bundle/client.ts:
```diff
@@ -78,9 +78,12 @@ export const createClient = (config: Config = {}): Client => {
     const requestInit: ReqInit = {
       redirect: 'follow',
       ...opts,
-      body: opts.serializedBody,
```
Isn't all that you want something like this?

```ts
body: opts.serializedBody !== undefined ? opts.serializedBody : opts.body,
```

franworks replied:
i thought about going the `opts.serializedBody ?? opts.body` route but decided to put it in a separate block for readability and clarity.

Body is already a property on the options object, no need to define or reassign if its not necessary. It was confusing for me to see body redefined in the client-next plugin like `body: options.body`. I had to ask myself if something was different - it wasn't.

Happy to change it if you prefer the one liner.

mrlubos reviewed (commented):
Please update the logic in all clients using `serializedBody` and then update snapshots with `pnpm --filter @test/openapi-ts test:update`

Commit:
fix(client-next): ensures plain/text body is sent with request

refs: #2553

Commit:
test(client): updates snapshots

refs: #2553

Commit:
fix(client-fetch): ensures valid serialized or unserialized body is sent with request

refs: #2553

Commit:
fix(client-fetch): ensures request body can be serialized when defined

refs: #2553

Commit:
fix(client-next): ensures valid serialized or unserialized body is sent with request

refs: #2553

Commit:
test(client): updates snapshots

refs: #2553

mrlubos committed:
Merge branch 'main' into fix/#2553-handle-plain-text-request-body

mrlubos commented on packages/openapi-ts/src/plugins/@hey-api/client-next/bundle/client.ts:
```diff
@@ -57,12 +57,15 @@ export const createClient = (config: Config = {}): Client => {
-    if (opts.serializedBody === undefined || opts.serializedBody === '') {
+    if (
+      opts.body === undefined ||
+      (opts.serializedBody !== undefined && opts.serializedBody === '')
```
The first check is redundant

```suggestion
      (opts.serializedBody === '')
```

franworks replied:
good point - thanks, I will update in both clients.

mrlubos commented on packages/openapi-ts/src/plugins/@hey-api/client-next/bundle/client.ts:
```diff
@@ -196,6 +201,25 @@ export const createClient = (config: Config = {}): Client => {
     };
   };
 
+  function getValidRequestBody(options: ResolvedRequestOptions) {
```
Why doesn't this function default to returning `undefined` instead of `null`?

franworks replied:
It could return undefined. The body will still be defined as null on the request.
Would you like me to update it?

franworks replied:
Actually, looking at how body is defined to begin with, undefined makes sense, even if the request will be null when sent. I will update.

Commit:
chore(client-fetch,client-nuxt): removes redundant delete header condition

Commit:
chore(client): updates snapshots

Commit:
chore(client-fetch,client-nuxt): updates getValidRequestBody to return undefined by default

Commit:
test(client): updates snapshots

Commit:
fix(client-angular)!: ensures valid serialized or unserialized body is sent with request

BREAKING CHANGE: options.serializedBody is undefined by default, and will only be assigned a value when options.bodySerializer is defined

refs: #2553

mrlubos committed:
Create brave-seahorses-dream.md

mrlubos committed:
Merge branch 'main' into fix/#2553-handle-plain-text-request-body

mrlubos reviewed (approved):
This looks good! I think we need to improve the Axios client as well, but happy to merge it
